carbonserver/render.go

```diff
@@ -355,20 +367,26 @@ func (listener *CarbonserverListener) prepareDataProto(ctx context.Context, logg
 		metricNames[i] = k
 		i++
 	}
-	expandedGlobs, err := listener.getExpandedGlobs(ctx, logger, time.Now(), metricNames)
-
-	if expandedGlobs == nil {
-		return fetchResponse{nil, contentType, 0, 0, 0, nil}, err
-	}
+	return metricNames
+}
 
+func getMetricGlobMapFromExpandedGlobs(expandedGlobs []globs) map[string]globs {
 	metricGlobMap := make(map[string]globs)
 	for _, expandedGlob := range expandedGlobs {
 		metricGlobMap[strings.ReplaceAll(expandedGlob.Name, "/", ".")] = expandedGlob
 	}
+	return metricGlobMap
+}
 
-	var metrics []string
+func (listener *CarbonserverListener) prepareDataStream(ctx context.Context, format responseFormat, targets map[timeRange][]target, metricGlobMap map[string]globs, responseChan chan response) {
+	defer close(responseChan)
 	for tr, ts := range targets {
 		for _, metric := range ts {
+			select {
+			case <-ctx.Done():
```

Comment:
should we need to check the context errors like [this](https://github.com/go-graphite/go-carbon/blob/quota/realtime-usage/carbonserver/carbonserver.go#L1248-L1253) here?

Reply:
How about having the check outside of prepareDataStream?
Like after reading from the channel, we can check the context errors and act accordingly.

Reply:
yep, no objection if it could give us insight of why the requests return early. 

but it seems there were no ctx.Err check in the non-stream logics as well? not a hard requirement for this pr.

Reply:
Yeah the older implementation didn't respect context deadline apparently. However, I added the check. Could you please take a look?

Reply:
yep. lgtm. tx.